cpp/src/join/semi_join.cu

```diff
@@ -36,26 +36,216 @@
+    // - https://github.com/rapidsai/cudf/pull/8277
+    if (rhs_has_nulls && compare_nulls == null_equality::UNEQUAL) {
+      // Gather all nullable columns at all levels from the right table.
+      auto const right_nullable_columns = [&] {
+        auto result = std::vector<column_view>{};
+        for (auto const& col : right_keys) {
+          accumulate_nullable_nested_columns(col, result);
+        }
+        return result;
+      }();
+
+      [[maybe_unused]] auto const [row_bitmask, tmp] =
+        cudf::detail::bitmask_and(table_view{right_nullable_columns}, stream);
+
+      // Insert only rows that do not have any nulls at any level.
+      map.insert_if(kv_it,
+                    kv_it + right_keys.num_rows(),
+                    thrust::counting_iterator<size_type>(0),  // stencil
+                    row_is_valid{static_cast<bitmask_type const*>(row_bitmask.data())},
+                    stream.value());
+    } else {
+      map.insert(kv_it, kv_it + right_keys.num_rows(), stream.value());
+    }
+  }
+
+  auto contained = rmm::device_uvector<bool>(left_keys.num_rows(), stream);
+
+  // Check contains for each row of the lhs table in the rhs table.
+  {
+    auto const hasher   = cudf::experimental::row::hash::row_hasher(left_keys, stream);
+    auto const d_hasher = hasher.device_hasher(nullate::DYNAMIC{lhs_has_nulls});
+
+    auto const kv_it = cudf::detail::make_counting_transform_iterator(
+      size_type{0},
+      make_pair_fn<lhs_index_type, decltype(d_hasher)>{d_hasher, map.get_empty_key_sentinel()});
+
+    auto const comparator =
+      cudf::experimental::row::equality::two_table_comparator(left_keys, right_keys, stream);
+
+    auto const do_check = [&](auto const& value_comp) {
+      auto const d_eqcomp = comparator.equal_to(
+        nullate::DYNAMIC{lhs_has_nulls || rhs_has_nulls}, compare_nulls, value_comp);
+
+      if (lhs_has_nulls && compare_nulls == null_equality::UNEQUAL) {
+        // Gather all nullable columns at all levels from the left table.
+        auto const left_nullable_columns = [&] {
+          auto result = std::vector<column_view>{};
+          for (auto const& col : left_keys) {
+            accumulate_nullable_nested_columns(col, result);
+          }
+          return result;
+        }();
+
+        [[maybe_unused]] auto const [row_bitmask, tmp] =
+          cudf::detail::bitmask_and(table_view{left_nullable_columns}, stream);
+
+        auto d_map           = map.get_device_view();
+        auto const pair_comp = pair_comparator_fn<decltype(d_eqcomp)>{d_eqcomp};
+
+        thrust::transform(
+          rmm::exec_policy(stream),
+          thrust::make_counting_iterator(0),
+          thrust::make_counting_iterator(left_keys.num_rows()),
+          contained.begin(),
+          check_contains<decltype(d_map), decltype(kv_it), decltype(pair_comp)>{
+            d_map, kv_it, static_cast<bitmask_type const*>(row_bitmask.data()), pair_comp});
```

Comment:
This is still incorrect due to using cooperative_group. I'm still trying to fix this.....

Reply:
@jrhemstad This is where I and @PointKernel think that we will need `pair_contains_if` so we can avoid computing row hash and row comparison altogether if there is any (nested) null in a row.

Reply:
`pair_contains_if` is the right way to get rid of the use of cuco device view APIs. Will work on it.